lib/bolt/application.rb

```diff
@@ -0,0 +1,619 @@
+      assert_project_file(config.project)
+
+      if config.project.modules.empty? && resolve
+        outputter.print_message(
+          "Project configuration file #{config.project.project_file} does not "\
+          "specify any module dependencies. Nothing to do."
+        )
+        return true
+      end
+
+      installer = Bolt::ModuleInstaller.new(outputter, pal)
+
+      installer.install(config.project.modules,
+                        config.project.puppetfile,
+                        config.project.managed_moduledir,
+                        config.module_install,
+                        force: force,
+                        resolve: resolve)
+    end
+
+    # Show modules available to the project.
+    #
+    # @return [Hash] A map of module directories to module definitions.
+    #
+    def module_show
+      pal.list_modules
+    end
+
+    # Convert a YAML plan to a Puppet language plan.
+    #
+    # @param plan [String] The plan to convert. Can be a plan name or a path.
+    # @return [String] The converted plan.
+    #
+    def plan_convert(plan)
+      pal.convert_plan(plan)
+    end
+
+    # Create a new project-level plan.
+    #
+    # @param name [String] The name of the new plan.
+    # @param puppet [Boolean] Create a Puppet language plan.
+    # @return [Boolean]
+    #
+    def plan_new(name, puppet: false)
+      Bolt::PlanCreator.validate_input(config.project, name)
+      Bolt::PlanCreator.create_plan(config.project.plans_path, name, puppet)
+    end
+
+    # Run a plan.
+    #
+    # @param plan [String] The plan to run.
+    # @param targets [Array[String], NilClass] The targets to pass to the plan.
+    # @param params [Hash] Parameters to pass to the plan.
+    # @return [Bolt::PlanResult]
+    #
+    def plan_run(plan, targets, params: {})
+      if targets && targets.any?
+        if params['nodes'] || params['targets']
+          key = params.include?('nodes') ? 'nodes' : 'targets'
+          raise Bolt::CLIError,
+                "A plan's '#{key}' parameter can be specified using the --#{key} option, but in that " \
+                "case it must not be specified as a separate #{key}=<value> parameter nor included " \
+                "in the JSON data passed in the --params option"
+        end
+
+        plan_params  = pal.get_plan_info(plan)['parameters']
+        target_param = plan_params.dig('targets', 'type') =~ /TargetSpec/
+        node_param   = plan_params.include?('nodes')
+
+        if node_param && target_param
+          msg = "Plan parameters include both 'nodes' and 'targets' with type 'TargetSpec', " \
+                "neither will populated with the value for --nodes or --targets."
+          Bolt::Logger.warn("nodes_targets_parameters", msg)
+        elsif node_param
+          params['nodes'] = targets.join(',')
+        elsif target_param
+          params['targets'] = targets.join(',')
+        end
+
+        inventory.get_targets(targets)
```

Comment:
This seems like it doesn't get stored anywhere - and actually, we already [get targets inside of run_plan function itself](https://github.com/puppetlabs/bolt/blob/main/bolt-modules/boltlib/lib/puppet/functions/run_plan.rb#L114) that ends up getting called, so that string targets from plans also work. I don't think this is needed.unless it's for validation.